### PR TITLE
Deprecate legacy cache + use cache position

### PR DESCRIPTION
**Pull request by zucchini-nlp**
# What does this PR do?

This PR deprecates legacy cache in all models that currently support Cache class. Also, these models now rely on cache position and update-causal-mask to get 4d attention mask. Tests are passing on my end, will activate slow tests with commit msg on PR later

### CODE REVIEW

File: src/transformers/models/dbrx/modeling_dbrx.py [hunk 1]
```diff
@@ -1119,6 +1119,10 @@ def forward(
         if use_cache and not isinstance(past_key_values, Cache):  # kept for BC (non `Cache` `past_key_values` inputs)
             return_legacy_cache = True
             past_key_values = DynamicCache.from_legacy_cache(past_key_values)
+            logger.warning_once(
+                "Using `past_key_values` as a tuple is deprecated and will be removed in v4.43. "
```
Comment (ArthurZucker) [inline src/transformers/models/dbrx/modeling_dbrx.py:1123]:
```suggestion
                "We detected that you are passing `past_key_values` as a tuple and this is deprecated and will be removed in v4.43. "
```

**Review by ArthurZucker — APPROVED** [review]
🔥 super nice! Thanks for porting this and adding the deprecation
I think in the futur we will want to use generation_config.cache_config.ma_length maybe instead of calling get_max_length() but anyways super good porting!

**Comment by zucchini-nlp** [comment]
> I think in the futur we will want to use generation_config.cache_config.ma_length maybe instead of calling get_max_length() 

Yes, sounds good if we can start adopting cache config for all cache related arguments

**Review by gante — APPROVED** [review]
💛

**Pull request merged**